### PR TITLE
[mod_callcenter] Refactoring the code - Untested

### PR DESCRIPTION
**Pull request by mochouinard**

### CODE REVIEW

File: src/mod/applications/mod_callcenter/mod_callcenter.c [hunk 1]
```diff
@@ -1759,67 +1800,108 @@ static switch_status_t playback_array(switch_core_session_t *session, const char
...
 	char *event_cc_header_count = NULL; 
 	char *event_cc_header_status = NULL;
 	switch_event_t *event = NULL;
 	char *sql = NULL;
+	int max_attempt = 0;
+	int current_attempt_count = 0;
+	cc_agent_status_t max_reached_status = CC_AGENT_STATUS_ON_BREAK;
+	agent_info_callback_t cbt;
+
+	cbt.found = SWITCH_TRUE;
```
Comment (mochouinard) [inline src/mod/applications/mod_callcenter/mod_callcenter.c:1819]:
```suggestion
	cbt.found = SWITCH_FALSE;
```

**Review by mochouinard — COMMENTED** [review]
Again, I haven't run the code... But except for this small error in this review, it looking good.

What do you think ?

**Comment by wmasilva** [comment]
sorry, i didn't see that you had done the changes here :( 
gonna fix the conflict.

**Comment by wmasilva** [comment]
testing it now.

The delay when not registered was a hardcode value of 5 seconds, what you think of setting the default value in DB to 5?

not_registered_delay_time INTEGER NOT NULL DEFAULT 5

**Comment by mochouinard** [comment]
5 sound good to me.  Long enough so it doesn't go in a quick loop, but not too long so an agent that become available that they don't wait too long for their phone to start ringing.

**Comment by wmasilva** [comment]
it segfaults when load the module.. i'm trying to figure out what is wrong:

```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/local/freeswitch/bin/freeswitch -nonat'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __strcmp_avx2 () at ../sysdeps/x86_64/multiarch/strcmp-avx2.S:736
736	../sysdeps/x86_64/multiarch/strcmp-avx2.S: No such file or directory.
[Current thread is 1 (Thread 0x7f248cb2b700 (LWP 11832))]
(gdb) bt
#0  0x00007f24c479ea27 in __strcmp_avx2 () at ../sysdeps/x86_64/multiarch/strcmp-avx2.S:736
#1  0x00007f24c4d50d9d in switch_xml_config_parse_event (event=<optimized out>, count=<optimized out>, reload=reload@entry=SWITCH_FALSE, instructions=instructions@entry=0x7f24ac2ddc98) at src/switch_xml_config.c:298
#2  0x00007f248c9bc327 in load_queue (queue_name=0x7f248c9d68db "support@default", request_agents=request_agents@entry=SWITCH_FALSE, request_tiers=request_tiers@entry=SWITCH_FALSE, x_queues_cfg=x_queues_cfg@entry=0x7f24ac0ccc60) at mod_callcenter.c:833
#3  0x00007f248c9c3453 in load_config () at mod_callcenter.c:1751
#4  0x00007f248c9c3453 in mod_callcenter_load (pool=0x7f24ac2bb358, module_interface=0x7f248cb2a648) at mod_callcenter.c:4433
#5  0x00007f248c9c3453 in mod_callcenter_load (module_interface=module_interface@entry=0x7f248cb2a648, pool=0x7f24ac2bb358) at mod_callcenter.c:4408
#6  0x00007f24c4cd86d2 in switch_loadable_module_load_file (new_module=<synthetic pointer>, global=<optimized out>, filename=0x55d46a00a780 "mod_callcenter", path=0x55d46a00a790 "/usr/local/freeswitch/mod/mod_callcenter.so") at src/switch_loadable_module.c:1751
#7  0x00007f24c4cd86d2 in switch_loadable_module_load_module_ex (dir=<optimized out>, fname=<optimized out>, runtime=SWITCH_TRUE, global=<optimized out>, err=0x7f248cb2a6c8, type=SWITCH_LOADABLE_MODULE_TYPE_COMMON, event_hash=0x0) at src/switch_loadable_module.c:1859
#8  0x00007f24c4cdaa76 in switch_loadable_module_load_module (dir=<optimized out>, fname=fname@entry=0x7f24ac078620 "mod_callcenter", runtime=runtime@entry=SWITCH_TRUE, err=err@entry=0x7f248cb2a6c8) at src/switch_loadable_module.c:1813
#9  0x00007f24a1f781f5 in load_function (session=<optimized out>, stream=0x7f248cb2ab90, cmd=0x7f24ac078620 "mod_callcenter") at mod_commands.c:2706
#10 0x00007f24a1f781f5 in load_function (cmd=0x7f24ac078620 "mod_callcenter", session=<optimized out>, stream=0x7f248cb2ab90) at mod_commands.c:2691
#11 0x00007f24c4cde4f5 in switch_api_execute (cmd=cmd@entry=0x7f24ac078640 "load", arg=arg@entry=0x7f24ac078645 "mod_callcenter ", session=session@entry=0x0, stream=stream@entry=0x7f248cb2ab90) at src/switch_loadable_module.c:3006
#12 0x00007f24c4c64238 in switch_console_execute (xcmd=<optimized out>, rec=0, istream=0x7f248cb2ab90) at src/switch_console.c:393
#13 0x00007f24c4c64380 in switch_console_process (xcmd=xcmd@entry=0x7f24ac0430d0 "load mod_callcenter ") at src/switch_console.c:321
#14 0x00007f24c4c6449c in console_thread (thread=<optimized out>, obj=<optimized out>) at src/switch_console.c:1077
#15 0x00007f24c4f920fc in dummy_worker (opaque=0x55d46a02abc8) at threadproc/unix/thread.c:151
#16 0x00007f24c4b8afa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#17 0x00007f24c47404cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
```

**Comment by mochouinard** [comment]
I just did a checkout from your branch, but the line don't match that backtrace...  Are you running a different version than what in your branch on github ?

**Comment by wmasilva** [comment]
sorry, it is the same branch, but i've local changes... just remove then.

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/local/freeswitch/bin/freeswitch -nonat'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __strcasecmp_l_avx () at ../sysdeps/x86_64/multiarch/strcmp-sse42.S:200
200	../sysdeps/x86_64/multiarch/strcmp-sse42.S: No such file or directory.
[Current thread is 1 (Thread 0x7fdce614f700 (LWP 22489))]
(gdb) bt
#0  0x00007fdd1d5c4bb5 in __strcasecmp_l_avx () at ../sysdeps/x86_64/multiarch/strcmp-sse42.S:200
#1  0x00007fdce57db548 in cc_agent_str2status (str=str@entry=0x0) at mod_callcenter.c:405
#2  0x00007fdce57dc894 in queue_set_config (queue=queue@entry=0x7fdcfc2f1c18) at mod_callcenter.c:649
#3  0x00007fdce57df308 in load_queue (queue_name=0x7fdce57f98db "support@default", request_agents=request_agents@entry=SWITCH_FALSE, request_tiers=request_tiers@entry=SWITCH_FALSE, x_queues_cfg=x_queues_cfg@entry=0x7fdcfc125810) at mod_callcenter.c:824
#4  0x00007fdce57e6473 in load_config () at mod_callcenter.c:1747
#5  0x00007fdce57e6473 in mod_callcenter_load (pool=0x7fdcfc2cf388, module_interface=0x7fdce614e648) at mod_callcenter.c:4429
#6  0x00007fdce57e6473 in mod_callcenter_load (module_interface=module_interface@entry=0x7fdce614e648, pool=0x7fdcfc2cf388) at mod_callcenter.c:4404
#7  0x00007fdd1dafd6d2 in switch_loadable_module_load_file (new_module=<synthetic pointer>, global=<optimized out>, filename=0x559eeed0a4c8 "mod_callcenter", path=0x559eeed0a4d8 "/usr/local/freeswitch/mod/mod_callcenter.so") at src/switch_loadable_module.c:1751
#8  0x00007fdd1dafd6d2 in switch_loadable_module_load_module_ex (dir=<optimized out>, fname=<optimized out>, runtime=SWITCH_TRUE, global=<optimized out>, err=0x7fdce614e6c8, type=SWITCH_LOADABLE_MODULE_TYPE_COMMON, event_hash=0x0) at src/switch_loadable_module.c:1859
#9  0x00007fdd1daffa76 in switch_loadable_module_load_module (dir=<optimized out>, fname=fname@entry=0x7fdcfc0e2110 "mod_callcenter", runtime=runtime@entry=SWITCH_TRUE, err=err@entry=0x7fdce614e6c8) at src/switch_loadable_module.c:1813
#10 0x00007fdcfb59f1f5 in load_function (session=<optimized out>, stream=0x7fdce614eb90, cmd=0x7fdcfc0e2110 "mod_callcenter") at mod_commands.c:2706
#11 0x00007fdcfb59f1f5 in load_function (cmd=0x7fdcfc0e2110 "mod_callcenter", session=<optimized out>, stream=0x7fdce614eb90) at mod_commands.c:2691
#12 0x00007fdd1db034f5 in switch_api_execute (cmd=cmd@entry=0x7fdcfc080300 "load", arg=arg@entry=0x7fdcfc080305 "mod_callcenter ", session=session@entry=0x0, stream=stream@entry=0x7fdce614eb90) at src/switch_loadable_module.c:3006
#13 0x00007fdd1da89238 in switch_console_execute (xcmd=<optimized out>, rec=0, istream=0x7fdce614eb90) at src/switch_console.c:393
#14 0x00007fdd1da89380 in switch_console_process (xcmd=xcmd@entry=0x7fdcfc0e01f0 "load mod_callcenter ") at src/switch_console.c:321
#15 0x00007fdd1da8949c in console_thread (thread=<optimized out>, obj=<optimized out>) at src/switch_console.c:1077
#16 0x00007fdd1ddb70fc in dummy_worker (opaque=0x559eeed29f98) at threadproc/unix/thread.c:151
#17 0x00007fdd1d9affa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#18 0x00007fdd1d5654cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by mochouinard** [comment]
I have to go, but it a little bit weird... It like agent-no-answer-status is null, but the config system is supposed when it empty to get the value cc_agent_status2str(CC_AGENT_STATUS_ON_BREAK)
With the backtrace, it would be to check if no_answer_status is really empty.  To check what value in your xml config for agent-no-answer-status.

**Comment by wmasilva** [comment]
Yes,  the variable is 0:
```
(gdb) frame 2
#2  0x00007fb5e1fbd894 in queue_set_config (queue=queue@entry=0x7fb6002de378) at mod_callcenter.c:649
649		queue->agent_no_answer_status = cc_agent_str2status(no_answer_status);
(gdb) print queue->agent_no_answer_status
$1 = CC_AGENT_STATUS_UNKNOWN
(gdb) print no_answer_status
$2 = 0x0
```

in the xml is not set.

**Comment by wmasilva** [comment]
i remove the from queue_set_config function the code:
```
        queue->agent_no_answer_status = cc_agent_str2status(no_answer_status);
	queue->agent_busy_status = cc_agent_str2status(busy_status);
	queue->agent_rejected_status = cc_agent_str2status(rejected_status);
	queue->agent_not_registered_status = cc_agent_str2status(not_registered_status);
```

this values are already initialize with:
`SWITCH_CONFIG_SET_ITEM(queue->config[i++], "agent-no-answer-status", SWITCH_CONFIG_STRING, 0, &queue->agent_no_answer_status, cc_agent_status2str(CC_AGENT_STATUS_ON_BREAK), &queue->config_str_pool, NULL, NULL);`

but i still got a segfault but now when parsing the xml config, i already force in the xml the parameters, but no luck. 
 I'm sill looking where the issue is, here is the full bt from this test. 

[bt_mod_callcenter.txt](https://github.com/wmasilva/freeswitch/files/4036599/bt_mod_callcenter.txt)

**Comment by wmasilva** [comment]
it looks that there is an issue reserving the queue memory, from the BT the pointer to the parameter is not accessible:

```
[Current thread is 1 (Thread 0x7f3cdbec3700 (LWP 12283))]
(gdb) bt
#0  0x00007f3d133376ca in __strcmp_avx2 () at ../sysdeps/x86_64/multiarch/strcmp-avx2.S:101
#1  0x00007f3d138e9d9d in switch_xml_config_parse_event (event=<optimized out>, count=<optimized out>, reload=reload@entry=SWITCH_FALSE, instructions=instructions@entry=0x7f3cf82a35d8) at src/switch_xml_config.c:298
#2  0x00007f3cdb553327 in load_queue (queue_name=0x7f3cdb56d8db "support@default", request_agents=request_agents@entry=SWITCH_FALSE, request_tiers=request_tiers@entry=SWITCH_FALSE, x_queues_cfg=x_queues_cfg@entry=0x7f3cf80d5bc0) at mod_callcenter.c:820
#3  0x00007f3cdb55a453 in load_config () at mod_callcenter.c:1738
#4  0x00007f3cdb55a453 in mod_callcenter_load (pool=0x7f3cf827ed68, module_interface=0x7f3cdbec2648) at mod_callcenter.c:4420
#5  0x00007f3cdb55a453 in mod_callcenter_load (module_interface=module_interface@entry=0x7f3cdbec2648, pool=0x7f3cf827ed68) at mod_callcenter.c:4395
#6  0x00007f3d138716d2 in switch_loadable_module_load_file (new_module=<synthetic pointer>, global=<optimized out>, filename=0x563528e1ba40 "mod_callcenter", path=0x563528e1ba50 "/usr/local/freeswitch/mod/mod_callcenter.so") at src/switch_loadable_module.c:1751
#7  0x00007f3d138716d2 in switch_loadable_module_load_module_ex (dir=<optimized out>, fname=<optimized out>, runtime=SWITCH_TRUE, global=<optimized out>, err=0x7f3cdbec26c8, type=SWITCH_LOADABLE_MODULE_TYPE_COMMON, event_hash=0x0) at src/switch_loadable_module.c:1859
#8  0x00007f3d13873a76 in switch_loadable_module_load_module (dir=<optimized out>, fname=fname@entry=0x7f3cf808a050 "mod_callcenter", runtime=runtime@entry=SWITCH_TRUE, err=err@entry=0x7f3cdbec26c8) at src/switch_loadable_module.c:1813
#9  0x00007f3d013191f5 in load_function (session=<optimized out>, stream=0x7f3cdbec2b90, cmd=0x7f3cf808a050 "mod_callcenter") at mod_commands.c:2706
#10 0x00007f3d013191f5 in load_function (cmd=0x7f3cf808a050 "mod_callcenter", session=<optimized out>, stream=0x7f3cdbec2b90) at mod_commands.c:2691
#11 0x00007f3d138774f5 in switch_api_execute (cmd=cmd@entry=0x7f3cf808dbd0 "load", arg=arg@entry=0x7f3cf808dbd5 "mod_callcenter ", session=session@entry=0x0, stream=stream@entry=0x7f3cdbec2b90) at src/switch_loadable_module.c:3006
#12 0x00007f3d137fd238 in switch_console_execute (xcmd=<optimized out>, rec=0, istream=0x7f3cdbec2b90) at src/switch_console.c:393
#13 0x00007f3d137fd380 in switch_console_process (xcmd=xcmd@entry=0x7f3cf8094340 "load mod_callcenter ") at src/switch_console.c:321
#14 0x00007f3d137fd49c in console_thread (thread=<optimized out>, obj=<optimized out>) at src/switch_console.c:1077
#15 0x00007f3d13b2b0fc in dummy_worker (opaque=0x563528e380f8) at threadproc/unix/thread.c:151
#16 0x00007f3d13723fa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#17 0x00007f3d132d94cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) frame 1
#1  0x00007f3d138e9d9d in switch_xml_config_parse_event (event=<optimized out>, count=<optimized out>, reload=reload@entry=SWITCH_FALSE, instructions=instructions@entry=0x7f3cf82a35d8) at src/switch_xml_config.c:298
298							if (!*dest || strcmp(*dest, newstring)) {
(gdb) print *dest
$1 = 0x7f3c <error: Cannot access memory at address 0x7f3c>
(gdb) print dest
$2 = (char **) 0x7f3cf82a3594
(gdb) print item->
data          defaultvalue  flags         function      helptext      key           ptr           syntax        type          
(gdb) print item->data 
$3 = (void *) 0x7f3cf82a80d8
(gdb) print item->defaultvalue 
$4 = (const void *) 0x7f3cdb55b186
(gdb) print *item->defaultvalue 
Attempt to dereference a generic pointer.
(gdb) print (char *)item->defaultvalue 
$5 = 0x7f3cdb55b186 "On Break"

```

From the code i can't see where it could be the error... the structure looks ok, the memory allocation also..  it must be something simple... :(

**Comment by wmasilva** [comment]
i think i got it, its a configuration item type, i guess that we cannot set SWITCH_CONFIG_STRING into a pointer of type cc_agent_status_t. 

i'm change the cc_queue struture to use char * and adjust the rest of the file.

**Comment by mochouinard** [comment]
Do not... It why I had the few lines at the bottom that you had just removed... It converted the String to the cc_agent_status_t

**Comment by wmasilva** [comment]
hum, yes, but it fails in the cc_agent_str2status because the the no_answer_status is NULL, 


I just finish and if i convert it to char *, no problem, and i can set the parameter from the xml. 
if it set to cc_agent_status_t i don't see how it converts the parameter from the xml..

**Comment by mochouinard** [comment]
The real question is why it return null... If null, it should return the default value we return with the cc_agent_status2str in the command parameters.

This is the comment you wrote earlier that actually converted the parameters from the xml :
i remove the from queue_set_config function the code:

        queue->agent_no_answer_status = cc_agent_str2status(no_answer_status);
	queue->agent_busy_status = cc_agent_str2status(busy_status);
	queue->agent_rejected_status = cc_agent_str2status(rejected_status);
	queue->agent_not_registered_status = cc_agent_str2status(not_registered_status);

**Comment by wmasilva** [comment]
i see, they shoud be in the function load_queue after switch_event_import_xml, no?

**Comment by mochouinard** [comment]
Correct, like they were before in my patch.  The only question is WHY is the xml code return NULL when it should return the value from the status2str if not set.
I'm working on other things right now, so I can't test it right away.

**Comment by wmasilva** [comment]
no problem, i'm trying all the possible options.

The initial segfault when queue->agent_no_answer_status = cc_agent_str2status(no_answer_status); 
It's because the no_answer_status is initialize to NULL, so the function cc_agent_str2status will segfault when it try to strcasecmp.

So this  code must go, and the init should be done by the default value.

I found the right way, setting the SWITCH_CONFIG_ENUM, see my commit. it loads ok, i'm testing now the behaviour.

thanks for the tips :)

**Comment by mochouinard** [comment]
Perfect, that even better !!!

**Comment by wmasilva** [comment]
Hi Marc, 
Test done with success! we can pr to FS.

**Comment by mochouinard** [comment]
Great !

**Comment by mochouinard** [comment]
I'm not 100% sure about alter table agents add not_registered_delay_time integer not null default 5...
I'm not sure if the default should be set at the DB level...

**Comment by mochouinard** [comment]
I'm not sure about not_registered_delay_time INTEGER NOT NULL DEFAULT 5,
Should we even allow it to be zero ? (I know it cause a super fast loopback).  And should we make it so if someone put zero, we force it to 5.  So we would keep it at 0 in the DB.
Or do we let to people stupidity and allow them to set it to 0 with it behavior ? ;)

**Comment by wmasilva** [comment]
yeahp, good question... i think is no harm if we force it to 5 if the value in DB is set to 0, it is the current behaviour and is more safe than set the value as default in db (one can put this value to 0 on the insert without notice). 
i will make the changes. 

There is also another question, in the config enum item, should we allow to set the value to "Unknown" for the agent status in the config?

**Comment by wmasilva** [comment]
Hi Marc, 
sorry for the delay, just commit the changes.

**Pull request merged**